text/0020-distingush_manual_vs_scripted_transforms.md

```diff
@@ -0,0 +1,312 @@
+
+## Proposal distinguish between "Manual" and "Scripted" transforms
+To me, the root of the problem here is we're being concise as the cost of clarity. it's clear that Qri is doing too many things in a single step, and allowing multiple things to affect the outcome of a `qri save`.  To solve this, we'll enforce some new rules and ratchet up some definitions. I want to start by adding formality to the the term "transform":
+
+### A "transform" is a forward transition from one dataset snapshot to another.
+* transforms must mutate one or more non-computed fields of a dataset
+* only one type of transform can be applied to any field per transform
+* transforms can use one or more types of mutations to determine the next snapshot
+
+A transform is the opposite of a history, moving forwards in snapshots instead of backwards. A history is _reproducible_ when you can start at the first snapshot and re-execute each mutation described in the next snapshot. By enforcing the mutually-exclusive mutations, each snapshot is a deterministic record of both state and _how to arrive at that state_ from the previous snapshot.
+
+#### Manual Transforms
+@ramfox had the lightbulb moment of conceiving of changes a user makes as a "human transform". The phrase "human transform" invites us to concieve of people manually changing things as a category of transformation, and that's exactly how we'll model it.
+
+Manual transforms work by providing values directly to Qri, such as with a `dataset.yaml` file:
+```yaml
+meta:
+  title: rational number series
+body: 1,2,3,4,5,6,7,8,9,10]
+```
+
+And a CLI command:
+```
+$ qri save --file=dataset.yaml me/rational_numbers
+saved dataset b5/rational_numbers
+```
+
+We can conceive of the above as a _manual transform_: a series of assignments in a single function call, with no calculations or business logic. The above example is telling Qri to do the following:
+
+```python
+def human_transform(ds):
+  ds.set_meta("title", "rational number series")
+  ds.set_body([1,2,3,4,5,6,7,8,9,10])
+```
+
+#### Scripted Transforms
+With a scripted transform, instead of making manual mutations, an algorithm _automates_ changes to fields of a dataset with programmatic instructions:
+
+```python
+def transform(ds, ctx):
+  ds.set_meta("title", "rational number series")
+  # use the range function to automate setting the dataset body
+  ds.set_body(range(1,11))
+```
+
+### Transform types are mutually exclusive
+Both types of transform are acting on the same components and fields of a dataset (`meta` and `body` components, a `title` field and body `rows`). To ensure reproducibility, we need a new rule: **only one type of transform can mutate a field between two snapshots**. 
```

Comment:
I would prefer to phrase this as "each type of transform must only mutate the dataset in a way that is composable with all other transforms". This is more future proof, as it eventually lets us weaken the current reality.

Reply:
lovely. I think both can/must be true. Will add.